### ISSUE TITLE
TF.data.dataset.cache(path) still uses memory despite the cache file path is given in tf2.8

### ISSUE THREAD

**Issue by yusenzhan** [tensorflow/tensorflow#55355]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
binary
- TensorFlow version (use command below):
tf2.8.0
- Python version:
python 3.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
CUDA v11.2. cuDNN v8.1
- GPU model and memory:
K80, gMem=16GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I came across a weird problem when I read TFrecords files from S3 through tf.dataset and cached them to my local path. Here is my reading code

    filenames=['s3s:path1', ''s3s:path2']
    dataset = tf.data.TFRecordDataset(filenames, compression_type="GZIP")
    parsed_dataset = (
        dataset.batch(batch_size, num_parallel_calls=tf.data.AUTOTUNE)
        .map(decode, num_parallel_calls=tf.data.AUTOTUNE)
        .cache(cache_file_path)
        .prefetch(tf.data.AUTOTUNE)
    )
It’s very strange that cache() still uses the internal memory which results in OOM. Here is the memory usage I printed via callback during training.

2022-03-08T22:19:40.154191003Z ...Training: end of batch 15700; got log keys: ['loss', 'copc', 'auc']
2022-03-08T22:19:40.159188560Z totalmemor: 59.958843GB
2022-03-08T22:19:40.159223737Z availablememory: 8.418320GB
2022-03-08T22:19:40.159250296Z usedmemory: 50.959393GB
2022-03-08T22:19:40.159257814Z percentof used memory: 86.000000
2022-03-08T22:19:40.159263710Z freememory:1.072124GB
2022-03-08T22:19:47.752077011Z Tue Mar  8 22:19:47 UTC 2022	job-submitter:	job run error: signal: killed

**Describe the expected behavior**
I have tested the code on TF2.3 which has no such an issue, but TF2.8 have such an OOM issue. It's supposed that when the path is given, cache(path) should use files to cache instead of memory.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by yusenzhan** [comment]
If need any details, please let me know. Thanks

**Comment by sushreebarsa** [comment]
@yusenzhan 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by sushreebarsa** [comment]
@yusenzhan 
Could you please respond  to the [comment](https://github.com/tensorflow/tensorflow/issues/55355#issuecomment-1077336476) above?
If you are using  this  
```
if cache: self.train_ds.cache()
```
Please change this to following as it will increase the performance
```
if cache: train_hr_ds = train_hr_ds.cache()
```
Thank you!

**Comment by sushreebarsa** [comment]
Closing this issue due to lack of recent activity. Please feel free to reopen the issue if you still have a concern. Thanks!

**Comment by DenisVorotyntsev** [comment]
I faced a similar problem in our pipelines. I trained models using tf 2.4.3, and memory consumption was fine. When testing tf 2.8.1 I faced an increase in memory usage and oom during the first epoch. After an investigation I found the problem - I had a manual processing function after `cache`, i.e.:

```
dataset = dataset.cache(path)
dataset = dataset.map(lambda x: some_mapping_f(x))
```

After moving the `some_mapping_f` logic to loss and metrics (luckily I could do that) problem was gone, the memory footprint became the same as in tf 2.4.3. 

Hope it will help.
